ui/components/app/snaps/snap-permissions-list/snap-permissions-list.js

```diff
@@ -1,52 +1,114 @@
+import {
+  getFilteredSnapPermissions,
+  getSnapName,
+} from '../../../../helpers/utils/util';
+import { getWeightedPermissions } from '../../../../helpers/utils/permission';
+import SnapPermissionAdapter from '../snap-permission-adapter';
 
 export default function SnapPermissionsList({
   snapId,
   snapName,
   permissions,
   connections,
   showOptions,
+  showAllPermissions,
+  onShowAllPermissions,
 }) {
   const t = useI18nContext();
-  const snapsMetadata = useSelector(getSnapsMetadata);
-  const permissionsToShow = {
-    ...permissions,
-    connection_permission: connections ?? {},
-  };
+
+  const combinedPermissions = useMemo(() => {
+    return { ...permissions, connection_permission: connections ?? {} };
+  }, [permissions, connections]);
+
   const targetSubjectsMetadata = useSelector((state) =>
     getMultipleTargetsSubjectMetadata(state, connections),
   );
 
+  const snapsMetadata = useSelector(getSnapsMetadata);
+
+  const [showAll, setShowAll] = useState(false);
+
+  if (showAllPermissions && !showAll) {
+    setShowAll(true);
+  }
+
+  const weightedPermissions = getWeightedPermissions({
+    t,
+    permissions: combinedPermissions,
+    subjectName: snapName,
+    getSubjectName: getSnapName(snapsMetadata),
+  });
+
+  if (
+    Object.keys(weightedPermissions).length <=
+      PermissionsAbstractionThreshold &&
+    !showAll
+  ) {
+    setShowAll(true);
+  }
+
+  const filteredPermissions = getFilteredSnapPermissions(
+    weightedPermissions,
+    PermissionWeightThreshold.snapInstall,
+    true,
+  );
+
+  const onShowAllPermissionsHandler = () => {
+    onShowAllPermissions();
+    setShowAll(true);
+  };
+
   return (
-    <Box className="snap-permissions-list">
-      {getWeightedPermissions({
-        t,
-        permissions: permissionsToShow,
-        subjectName: snapName,
-        getSubjectName: getSnapName(snapsMetadata),
-      }).map((permission, index) => (
-        <SnapPermissionCell
-          snapId={snapId}
-          showOptions={showOptions}
-          connectionSubjectMetadata={
-            targetSubjectsMetadata[permission.connection]
-          }
-          permission={permission}
-          index={index}
-          key={`permissionCellDisplay_${snapId}_${index}`}
-        />
-      ))}
+    <Box display={Display.Flex} flexDirection={FlexDirection.Column}>
+      <Box className="snap-permissions-list">
+        {showAll ? (
+          <SnapPermissionAdapter
+            permissions={weightedPermissions}
+            snapId={snapId}
+            snapName={snapName}
+            showOptions={showOptions}
+            targetSubjectsMetadata={targetSubjectsMetadata}
+          />
+        ) : (
+          <SnapPermissionAdapter
+            permissions={filteredPermissions}
+            snapId={snapId}
+            snapName={snapName}
+            showOptions={showOptions}
+            targetSubjectsMetadata={targetSubjectsMetadata}
+          />
+        )}
```

Comment:
I wish to avoid having to use mapping and similar array operations or logical complexities within the JSX template. 

Extracting it to a utility function at the top level of the component also adds more code and make it more difficult to maintain the component. I would prefer to keep the wrapper if you would agree.

Reply:
At the very least wouldn't this be simpler?
```suggestion
          <SnapPermissionAdapter
            permissions={showAll ? weightedPermissions : filteredPermissions}
            snapId={snapId}
            snapName={snapName}
            showOptions={showOptions}
            targetSubjectsMetadata={targetSubjectsMetadata}
          />
```

Reply:
Yea, that can be done. Updated.